src/language-js/print/type-annotation.js

```diff
@@ -42,6 +42,7 @@ function shouldHugType(node) {
       (node) =>
         node.type === "ObjectTypeAnnotation" ||
         node.type === "TSTypeLiteral" ||
+        node.type === "TSUnionType" ||
```

Comment:
Should we skip printing `|` here, 

https://github.com/prettier/prettier/blob/37fb53acf33a20379a93bfecea8242c03ce0a4fc/src/language-js/print/type-annotation.js#L123

when union type only have one `types`, instead of changing this?

Reply:
As I tested, if I add

```js
  if (node.types.length === 1) {
    return print(["types", 0]);
  }
```

to begin of this function, they all format correctly(though there is a comment print issue to fix).

Reply:
Yes, your way of `node.types` checking looks much better. And I'm reached this issue with comments too, it occurs after formatting already formatted code, right?

Reply:
I'm not sure what happened, should be something wrong in `willPrintOwnComments`.

If we choose go this approach, the `parent` check part need a big update too. I suggest you test both `typescript` and `flow` parser, they should behave the same. They are not now, because the AST is different.

Reply:
Since there is no single type wrapper in flow, we also need update `canAttachComment`, otherwise the comments will attch to different nodes.

Reply:
Maybe strip these nodes with single type in `src/language-js/parse/postprocess/index.js` is a simplier way. 

Otherwise this seems be hard to fix, we also need update all places checking `TSUnionType`...

Reply:
Hi @fisker! Sorry for mention, but I'm really stuck on this issue. As I can see the stable version also has some troubles with comments inconsistency after re-formatting (like [this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAZgSwE4DOM2EeAtgIYww5QDmAOrAJ4AOcGFGAvExhgDIMACj4YGMVACoMARgxTMYicDH9+AazjMkGAG4UANgFc4ajAF8xASiZN0GAnEhQAJiXJUa9JjDYcubgxpOQVUJlUoTW1dAxM4AG4mCwSQABoQCFYaaAJkUAo8PAgAdwAFQoQ8lCMSimY8jIAjPAowLRgAZQoyOAAZWjhkLCMnZtb2uC7WNto6ZBg8Uwy4Mia4V1cNvop6Ywo6OAAxUkpqOeQQCmMYCHSQAAsYMkMAdQeceAIZsDhOqs+OD0n2YlzABEaIFoTjwMDKrTolGGozgGQAVgQAB6dOaGOAARWMEHgyMMYxAM0IcDwlz87AIYDwOGy91YTNgrxwrhgD2QAA4AAwZNkQJyvVqsS5suAwvRDDIARyJ8HhWWqVwIAFooHANht7ng4Er8HB4QckUgRmTUSAnGQcAsljaCLi4ABBahMpo3U3UgY60nkl30PGE4lDS0ojIwChNTnc3lIABM0daOEMcwAwhAyBaQDKAKz3YxOAAqseqVvJelMAEk3AguozmTA3W5On48YG4BYLEA) and [this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAZgSwE4DOM2EeAtgIYww5QDmAOrAJ4AOcGFGAvBkxhlQAqPlAEBGfhiGYAPhmABrOMyQYAbhQA2AVzgBfDPKUq1m3QaMLlqjdr2HjNs-ctPTdi-qZN0GAnCQUAAmJORUNPRMMGwcXLzCohJSMlLGGM6eehjeYlbAGR7m2bkC6ZnFHKX5hbaVOWkKtS4WOQDcIAA0IBCsNNAEyKAUeHgQAO4ACiMIgyja4xTMg90ARngUYMowAMoUZHAAMrRwyFjaAWsbW3C7rJu0dMgweHrdcGSrcMHB34cU9B0FDocAAYqRKNRHsgQBQdDAIF0QAALGBkLQAdWROHgBHuYDgO1mOJw6hxzBhYAIKxAtACeBgkw2dEoZwucG6ACsCAAPHaPLRwACKOgg8DZWkuIHuhDgeBhMXYBDAeBwfSRrFVsAxOGCMGRyAAHAAGbqaiABDEbVgwzVwenqU7dACOovgTN6c1hBAAtFA4N9vki8HBXfg4EzgaykOdJRyQAEyDhnq94wQBXAAILUVWreERuXHf0SqXp+iCkVi04x9ndGAUVY6vUGpAAJjrGxwWkeAGEIGRoyB7QBWJE6AIAFQbc1jUvUegAkiEELsVWqYJmQjsYoKSwZ9EA)). Could you please advice me what to do to solve this problem?

Reply:
Okay, let's pick the easy path first,

add

```js
      case "TSUnionType":
        if (node.types.length === 1) {
          return node.types[0]
        }
        break;
```

to

https://github.com/prettier/prettier/blob/f602be5a478a5bcdca870b31936dc873ecae1120/src/language-js/parse/postprocess/index.js#L63

the problem should be solved.

(I'm on cellphone, haven't test.)